### ISSUE TITLE
New device: Novation A Station

### ISSUE THREAD

**Issue by christofmuc** [christofmuc/KnobKraft-orm#267]
Now seems to be working, I'm super excited, this is great software!

```
#   Copyright (c) 2021 Christof Ruch. All rights reserved.
#   Novation Mini/Ultranova Adaptation version 0.6 by Cedric Tessier, 2022.
#   Dual licensed: Distributed under Affero GPL license by default, an MIT license is available for purchase
#
# based on my own (partial) reverse engineering of Novation UltraNova librarian.
#
# Attempt to refactor for A Station by Rothchild 17-09-23
#
import hashlib
from typing import List

import testing

NAME_OFFSET = 15
NAME_LEN = 16
NAME_SPECIALCHARSET = ' !"#$%&\'()*+,-./:;<=>?@[]^_`{|}'


def setupHelp():
    return '''\
Make sure the synth is not in protected mode (Global -> Protect -> Off).
'''


def wrapSysex(data):
    return [0xf0,  # Sysex message
            0x00, 0x20, 0x29, 0x01, # Novation
            #0x40,  # Product type?
            0x40,  # Device Type (1 = Synth)?
            0x7f   # ?
            ] + data + [0xf7]  # EOX


def name():
    return 'Novation AStaton'


def isDefaultName(patchName):
    return patchName == 'Init Patch'


def createDeviceDetectMessage(channel):
    return wrapSysex([0x40, 0x00, 0x00, 0x00, 0x00, 0x00])


def deviceDetectWaitMilliseconds():
    return 200


def generalMessageDelay():
    return 100


def needsChannelSpecificDetection():
    return False


def channelIfValidDeviceResponse(message):
    if (len(message) > 8
            and message[0] == 0xf0  # Sysex
            and message[1] == 0x00  # Novation
            and message[2] == 0x20  # Novation ID1
            and message[3] == 0x29  # Novation ID2
            and message[4] == 0x01  # Message Type ?
            and message[5] == 0x40  # Device Type
            and message[6] == 0x7f  # Channel
            and message[7] == 0x00):
        # TODO: message[11] is protection status
        # Extract midi channel from message
        return message[13] & 0xf
    return -1


def createEditBufferRequest(channel):
    return wrapSysex([0x40, 0x21, 0x00, 0x00, 0x00, 0x00])


def isEditBufferDump(message):
    return (len(message) > 13
            and message[0] == 0xf0  # Sysex
            and message[1] == 0x00  # Novation
            and message[2] == 0x20  # Novation ID1
            and message[3] == 0x29  # Novation ID2
            and message[4] == 0x01
            and message[5] == 0x40
            and message[6] == 0x7f
            and message[7] == 0x00  # ?
            #and message[8] == 0x00  # ?
            #and message[9] == 0x10  # ?
            #and message[10] == 0x00  # ?
            and message[11] == 0x00  # Bank number
            and message[12] == 0x00  # Slot number
            )


def numberOfBanks():
    return 4


def numberOfPatchesPerBank():
    return 100


def createProgramDumpRequest(channel, patchNo):
    bank = patchNo // numberOfPatchesPerBank()
    program = patchNo % numberOfPatchesPerBank()
    return wrapSysex([0x41, 0x00, 0x00, 0x00, bank+1, program])


def isSingleProgramDump(message):
    return (len(message) > 13
            and message[0] == 0xf0
            and message[1] == 0x00  # Novation
            and message[2] == 0x20  # Novation ID1
            and message[3] == 0x29  # Novation ID2
            and message[4] == 0x01
            and message[5] == 0x40
            and message[6] == 0x7f
            and message[7] in [0x00, 0x01]
            #and message[8] == 0x00  # ?
            #and message[9] == 0x10  # ?
            #and message[10] == 0x00  # ?
            and 0x00 <= message[11] <= numberOfBanks()  # Bank number [1-4] (can be 0 for an unknown reason...)
            and 0x00 <= message[12] < numberOfPatchesPerBank()  # Slot number [0-127]
            and not message[7] == message[11] == 0  # make sure this is NOT an edit buffer
            )


def numberFromDump(message):
    if not isSingleProgramDump(message):
        return 0
    bank = message[11]
    bank = bank - 1 if bank > 0 else 0  # some program patch have bank == 0
    program = message[12]
    return bank * numberOfPatchesPerBank() + program


def nameFromDump(message):
    if isSingleProgramDump(message) or isEditBufferDump(message):
        name = ''.join([chr(x) for x in message[NAME_OFFSET:NAME_OFFSET+NAME_LEN]])
        return name.strip()
    return 'Invalid'


def convertToEditBuffer(channel, message):
    if isEditBufferDump(message):
        return message[:8] + [0] + message[9:]
    elif isSingleProgramDump(message):
        # Clear Bank / Slot from message
        return message[:7] + [0, 0] + message[9:11] + [0, 0] + message[13:]
    raise Exception("Neither edit buffer nor program dump - can't be converted")


def convertToProgramDump(channel, message, program_number):
    bank = program_number // numberOfPatchesPerBank()
    program = program_number % numberOfPatchesPerBank()
    if isEditBufferDump(message) or isSingleProgramDump(message):
        return message[:7] + [1, 0] + message[9:11] + [bank+1, program] + message[13:]
    raise Exception("Neither edit buffer nor program dump - can't be converted")


def renamePatch(message, new_name):
    if isSingleProgramDump(message) or isEditBufferDump(message):
        # name is alpha-num + special chars (replace everything else by '_' and pad with spaces)
        clean_name = new_name.strip()[:NAME_LEN].ljust(NAME_LEN, ' ')
        valid_name = [ord(x) if x.isalnum() or x in NAME_SPECIALCHARSET else ord('_') for x in clean_name]
        return message[:NAME_OFFSET] + valid_name + message[NAME_OFFSET+NAME_LEN:]
    raise Exception("Neither edit buffer nor program dump can't be converted")


def friendlyBankName(bank_number):
    bank = bank_number + 1
    return {1: 'A', 2: 'B', 3: 'C', 4: 'D'}.get(bank, '')


# TODO: current implementation is broken as 'program' is never above 128...
#def friendlyProgramName(program):
#    return "%s%d" % (friendlyBankName(program // numberOfPatchesPerBank()), program % numberOfPatchesPerBank())


def calculateFingerprint(message):
    # Skip sysex, some header bytes and patch name
    data = message[1:7] + message[NAME_OFFSET+NAME_LEN:-1]
    return hashlib.md5(bytearray(data)).hexdigest()


def make_test_data():

    def make_patches(test_data: testing.TestData) -> List[testing.ProgramTestData]:
        # Additional test for rename
        patch = test_data.all_messages[0]
        same = renamePatch(patch, "Cr4zy Name°$    overflow")
        assert nameFromDump(same) == "Cr4zy Name_$"

        yield testing.ProgramTestData(message=test_data.all_messages[0], name="Poppy", rename_name="Paraver", number=35)

    return testing.TestData(sysex=R"testData/Novation_Ultranova_poppy.syx", program_generator=make_patches, friendly_bank_name=(2, 'C'))
```

_Originally posted by @thechildofroth in https://github.com/christofmuc/KnobKraft-orm/discussions/265#discussioncomment-7050364_

**Comment by thechildofroth** [comment]
On further testing there seems to be an issue with bank selection, I wonder if it's something to do with the message numbers that are assigned in 'def isSingleProgramDump'? (or if I don't need the +1 in friendlyBankName?)

From the specs (https://fael-downloads-prod.focusrite.com/customer/prod/downloads/A-Station%20User%20Guide%20-%20EN.pdf)

PROGRAM DUMP REQUEST
1. F0h SYSEX START
2. 00h NOVATION ID 1
3. 20h NOVATION ID 2
4. 29h NOVATION ID 3
5. 01h DEVICE TYPE
6. 40h A-STATION
7. SyCh SYSEX CHANNEL 7Fh or current receive channel
8. 41h MESSAGE TYPE Program dump request
9. C CONTROL BYTE Reply will copy this value for C
10. Vv SOFTWARE VERSION Don't care
11. Vi VERSION INCREMENT Don't care
12. B PROGRAM BANK 1-4 : don't care if C=0
13. P PROGRAM NUMBER 0-99
14. F7h END OF EXCLUSIVE

**Comment by christofmuc** [comment]
What are you experiencing? Do you download bank 1 as bank 0, or bank 1 as bank 2? The documentation suggests the bank is 1-based, so probably one needs to be added.

I usually program the adaptations so the programs just start at 0 and go up, and the next bank immediately starts after the previous bank. E.g. if a synth has 4 banks of 64 pages, I do 0..63 for bank 1, 64..127 for bank 2 etc.
Then of course you have to make sure to calculate the bank properly and add 1 in case it is one based.

**Comment by christofmuc** [comment]
Sometimes the documentation is plain wrong and 0 is a valid bank nunber...

**Comment by christofmuc** [comment]
First version released with 2.0.8
